### PR TITLE
feat(sequencer): Implement new create sequencer design

### LINKED ISSUE #979: Update sequencer registration process with new design
As discussed in this https://www.notion.so/dymension/ADR-x-Rollapp-registration-process-07f3d4f2ceed4ae9ab44f609abb6b1d1?pm=c and this https://www.notion.so/dymension/ADR-x-Implementation-of-RollApp-Registration-d548ea1d12ba4cf1bf2ae68fe9354332?pvs=4 ADRs, make changes to the existing implementation of the sequencer registration process.

### PR DESCRIPTION
**Pull request by zale144**
## Description

----

Closes #979 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/rollapp/keeper/msg_server_update_rollapp.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+func (k msgServer) UpdateRollappInformation(goCtx context.Context, msg *types.MsgUpdateRollappInformation) (*types.MsgUpdateRollappInformationResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
```
Comment (mtsitrin) [inline x/rollapp/keeper/msg_server_update_rollapp.go:13]: check msg not nit

File: x/rollapp/keeper/rollapp.go [hunk 2]
```diff
@@ -20,18 +21,20 @@ func (k Keeper) RegisterRollapp(ctx sdk.Context, rollapp types.Rollapp) error {
...
-		return fmt.Errorf("check if bech32 prefix taken: %w", err)
+		return err
 	}
 
 	creator, _ := sdk.AccAddressFromBech32(rollapp.Creator)
 	registrationFee := sdk.NewCoins(k.RegistrationFee(ctx))
 
-	if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, creator, types.ModuleName, registrationFee); err != nil {
-		return errorsmod.Wrap(types.ErrFeePayment, err.Error())
+	if registrationFee.IsAllPositive() {
```
Comment (mtsitrin) [inline x/rollapp/keeper/rollapp.go:34]:
```suggestion
	if !registrationFee.IsZero() {
```

File: x/rollapp/keeper/rollapp.go [hunk 3]
```diff
@@ -48,6 +51,120 @@ func (k Keeper) RegisterRollapp(ctx sdk.Context, rollapp types.Rollapp) error {
 	return nil
 }
 
+func (k Keeper) UpdateRollapp(ctx sdk.Context, update *types.UpdateRollappInformation) error {
```
Comment (mtsitrin) [inline x/rollapp/keeper/rollapp.go:54]:
```suggestion
func (k Keeper) UpdateRollapp(ctx sdk.Context, update types.UpdateRollappInformation) error {
```

File: x/rollapp/keeper/rollapp.go [hunk 4]
```diff
@@ -48,6 +51,120 @@ func (k Keeper) RegisterRollapp(ctx sdk.Context, rollapp types.Rollapp) error {
...
+
+	return nil
+}
+
+func (k Keeper) checkIfCanUpdateRollapp(ctx sdk.Context, update *types.UpdateRollappInformation) (current types.Rollapp, err error) {
+	var found bool
+	current, found = k.GetRollapp(ctx, update.RollappId)
+	if !found {
+		err = errRollappNotFound
+		return
```
Comment (mtsitrin) [inline x/rollapp/keeper/rollapp.go:79]: just `return errRollappNotFound` instead of using named out args

File: x/rollapp/keeper/rollapp.go [hunk 5]
```diff
@@ -48,6 +51,120 @@ func (k Keeper) RegisterRollapp(ctx sdk.Context, rollapp types.Rollapp) error {
...
+
+	return
+}
+
+func (k Keeper) checkIfCanUpdateInitialSequencer(ctx sdk.Context, currentInitSeq, updateInitSeq string, hasState bool) (string, error) {
+	if updateInitSeq == "" {
+		return currentInitSeq, nil
+	}
+
+	if currentInitSeq != "" {
```
Comment (mtsitrin) [inline x/rollapp/keeper/rollapp.go:128]: why is that?

File: x/rollapp/keeper/rollapp.go [hunk 6]
```diff
@@ -48,6 +51,120 @@ func (k Keeper) RegisterRollapp(ctx sdk.Context, rollapp types.Rollapp) error {
...
+		err = sdkerrors.ErrUnauthorized
+		return
+	}
+
+	if current.Frozen {
+		err = types.ErrRollappFrozen
+		return
+	}
+
+	_, hasState := k.GetLatestStateInfoIndex(ctx, update.RollappId)
```
Comment (mtsitrin) [inline x/rollapp/keeper/rollapp.go:92]: probably should check if some update requested to immutable field
and return if `hasState`

File: x/sequencer/types/errors.go [hunk 7]
```diff
@@ -27,4 +23,5 @@ var (
 	ErrInvalidType              = errorsmod.Register(ModuleName, 1016, "invalid type")
 	ErrUnknownRequest           = errorsmod.Register(ModuleName, 1017, "unknown request")
 	ErrInvalidRequest           = errorsmod.Register(ModuleName, 1018, "invalid request")
+	ErrSequencerJailed          = errorsmod.Register(ModuleName, 1019, "sequencer is jailed")
```
Comment (danwt) [inline x/sequencer/types/errors.go:26]: errorsmod.Wrap(gerrc.ErrFailedPrecondition, "sequencer is jailed")

File: x/sequencer/types/metadata.go [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+		Rpcs:           d2.Rpcs,
+		EvmRpcs:        d2.EvmRpcs,
+		RestApiUrl:     d2.RestApiUrl,
+		ExplorerUrl:    d2.ExplorerUrl,
+		GenesisUrls:    d2.GenesisUrls,
+		ContactDetails: d2.ContactDetails,
+		ExtraData:      d2.ExtraData,
+		Snapshots:      d2.Snapshots,
+		GasPrice:       d2.GasPrice,
+	}
```
Comment (danwt) [inline x/sequencer/types/metadata.go:32]: so you cant update p2p or rpc or evm or rest api or explorer or genesis or snapshots or gas price?

File: x/sequencer/types/msg_create_sequencer.go [hunk 9]
```diff
@@ -72,27 +64,28 @@ func (msg *MsgCreateSequencer) ValidateBasic() error {
...
 
-		_, err = edwards.ParsePubKey(edwards.Edwards(), pk.Bytes())
-		// err means the pubkey validation failed
-		if err != nil {
-			return errorsmod.Wrapf(ErrInvalidPubKey, "%s", err)
-		}
+	// cast to cryptotypes.PubKey type
+	pk, ok := msg.DymintPubKey.GetCachedValue().(cryptotypes.PubKey)
+	if !ok {
+		return errorsmod.Wrapf(ErrInvalidType, "Expecting cryptotypes.PubKey, got %T", pk)
```
Comment (danwt) [inline x/sequencer/types/msg_create_sequencer.go:79]: no caps in error message

**Review by danwt — CHANGES_REQUESTED** [review]
Ditto some of what I said here https://github.com/dymensionxyz/dymension/pull/980#pullrequestreview-2191366741
Especially wrt. proto

File: proto/dymensionxyz/dymension/sequencer/metadata.proto [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  // genesis URLs
+  repeated string genesis_urls = 8;
+  // contact details
+  ContactDetails contact_details = 9;
+  // json dump the sequencer can add (limited by size)
+  bytes extra_data = 10;
+  // snapshots of the sequencer
+  repeated SnapshotInfo snapshots = 11;
+  // gas_price defines the value for each gas unit
+  string gas_price = 12 [(gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int"];
```
Comment (mtsitrin) [inline proto/dymensionxyz/dymension/sequencer/metadata.proto:33]: @omritoptix I think this unneeded.
Better way is to have **known** gas price is to put it on-chain

File: x/sequencer/keeper/msg_server_update_sequencer.go [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	rollapp, found := k.rollappKeeper.GetRollapp(ctx, msg.RollappId)
+	if !found {
+		return nil, types.ErrUnknownRollappID
+	}
+
+	if rollapp.Frozen {
+		return nil, types.ErrRollappJailed
+	}
```
[the remark this answers is not in the document]
Comment (zale144) [reply x/sequencer/keeper/msg_server_update_sequencer.go:37]: if everyone agrees

File: proto/dymensionxyz/dymension/sequencer/metadata.proto (hunk 10 above)
Comment (zale144) [reply proto/dymensionxyz/dymension/sequencer/metadata.proto:33]: this is for UX only I believe

File: x/sequencer/keeper/msg_server_update_sequencer.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (danwt) [reply x/sequencer/keeper/msg_server_update_sequencer.go:37]: @omritoptix ?

File: x/sequencer/types/metadata.go (hunk 8 above)
Comment (danwt) [reply x/sequencer/types/metadata.go:33]: @zale144 ?

File: proto/dymensionxyz/dymension/sequencer/metadata.proto [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+option go_package = "github.com/dymensionxyz/dymension/v3/x/sequencer/types";
+
+import "gogoproto/gogo.proto";
+
+// Metadata defines rollapp/sequencer extra information.
+message SequencerMetadata {
+  // moniker defines a human-readable name for the sequencer.
+  string moniker = 1;
+  // details define other optional details.
+  string details = 2;
```
[the remark this answers is not in the document]
Comment (danwt) [reply proto/dymensionxyz/dymension/sequencer/metadata.proto:13]: @zale144 @srene what is this field?

File: x/sequencer/keeper/msg_server_update_sequencer.go [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+	sequencer, found := k.GetSequencer(ctx, msg.Creator)
+	if !found {
+		return nil, types.ErrUnknownSequencer
+	}
+
+	if sequencer.Jailed {
+		return nil, types.ErrSequencerJailed
+	}
```
[the remark this answers is not in the document]
Comment (danwt) [reply x/sequencer/keeper/msg_server_update_sequencer.go:28]: ?

File: app/upgrades/v4/upgrade.go [hunk 14]
```diff
@@ -152,6 +163,37 @@ func ConvertOldRollappToNew(oldRollapp types.Rollapp) rollapptypes.Rollapp {
...
+			RestApiUrl:  "",         // TODO
+			ExplorerUrl: "",         // TODO
+			GenesisUrls: []string{}, // TODO
+			ContactDetails: &sequencertypes.ContactDetails{ // TODO
+				Website:  "", // TODO
+				Telegram: "", // TODO
+				X:        "", // TODO
+			}, // TODO
+			ExtraData: nil,                              // TODO
+			Snapshots: []*sequencertypes.SnapshotInfo{}, // TODO
```
[the remark this answers is not in the document]
Comment (zale144) [reply app/upgrades/v4/upgrade.go:191]: @danwt migrating old rollapps to the new model will require some additional info from the existing rollapps (on mainnet), for which there is an open [issue](https://github.com/dymensionxyz/dymension/issues/987)

**Pull request merged**